mlir_graphblas/engine.py

```diff
@@ -762,6 +742,14 @@ def add(
 
         function_names: List[str] = []
         mlir_ast = parse_mlir_string(mlir_text)
+        # Find the inner-most Module
```

Comment:
I think this can cause us to ignore functions, e.g. `function_1` would get ignored in this MLIR code
```
module  {
  func @function_1() {
    return
  }
  module  {
    func @function_2() {
      return
    }
  }
}
```

Perhaps a better approach might be to traverse the original `mlir_ast` and 
- append onto `mlir_functions` when we see a function
- recursively traverse when we see a module.

There are also other weird edge cases that we might miss. For example, we'd miss one of the modules in this code (which passes through `mlir-opt` with no errors):
```
module  {
  func @function_1() {
    module  {
      func @function_weird() {
        return
      }
    }
    return
  }
  module  {
    func @function_2() {
      return
    }
  }
}
```
I'm not sure if we care about this last example. I think we can safely ignore this case for now.